plugins/modules/archive.py

```diff
@@ -575,16 +614,34 @@ def open(self):
             self.module.fail_json(msg="%s is not a valid archive format" % self.format)
 
     def _add(self, path, archive_name):
-        def py27_filter(tarinfo):
-            return None if matches_exclusion_patterns(tarinfo.name, self.exclusion_patterns) else tarinfo
+        def filter(tarinfo):
+            # type: (tarfile.TarInfo) -> tarfile.TarInfo | None
+            if matches_exclusion_patterns(tarinfo.name, self.exclusion_patterns):
+                return None
+
+            if self.reproducible_tar:
+                # Remove unused backref that prevents copy
+                if hasattr(tarinfo, "tarfile"):
+                    delattr(tarinfo, "tarfile")
+
+                if tarinfo.isdir():
+                    mode = 0o40000 | 0o755
+                else:
+                    mode = 0o100000 | (0o755 if tarinfo.mode & 0o100 else 0o644)
+
+                # Copy tarfile while reducing metadata
+                return tarinfo.replace(
+                    mtime=self._reproducible_mtime(),
+                    mode=mode,
+                    uid=0,
+                    gid=0,
+                    uname="",
+                    gname="",
+                )
```

Comment:
Hmm, this is a very special interpretation of 'reproducible tarfile' IMO. Also this is potentially dangerous since files that are protected (only readable by specific users/groups) before archiving are suddenly publicly readable after extraction.

Maybe it would be better to make the level of reproducibility configurable? On the other hand, that would make the interface also pretty complicated.

I guess this needs to be discussed first. Maybe create a thread in https://forum.ansible.com/c/project/collection-development/27 for that?

Reply:
It's not intended to be particularly special. I compare it to `tar` in the summary.

https://reproducible-builds.org/docs/archives/
https://www.gnu.org/software/tar/manual/html_section/Reproducibility.html
https://github.com/drivendataorg/repro-tarfile

Reply:
I'll make a thread when I have more time, I'll just leave some initial notes here.

* The danger is on extract of course.
* Unless you are extracting into a path that's already readable, it won't become readable:
   ```
    root@dev-master-1:~# tar -xzf coreutils-0.0.27-x86_64-unknown-linux-musl.tar.gz
    root@dev-master-1:~# sudo -u nobody ls /root/coreutils-0.0.27-x86_64-unknown-linux-musl
    ls: cannot access '/root/coreutils-0.0.27-x86_64-unknown-linux-musl': Permission denied
    root@dev-master-1:~# sudo -u nobody ls /root/coreutils-0.0.27-x86_64-unknown-linux-musl/LICENSE
    ls: cannot access '/root/coreutils-0.0.27-x86_64-unknown-linux-musl/LICENSE': Permission denied
   ```
* Extracting files as root (common with ansible) with odd source attributes (localhost ansible is not running as root), which was what was happening without this for me, is also potentially dangerous - UIDcollision leads to write privilege escalation.